Make <input type=checkbox switch> zoomable on macOS

annevk opened the pull request:
#### 9b4ea113ce2d829f4e4521771621fc0eb82853d1
<pre>
Make &lt;input type=checkbox switch&gt; zoomable on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=267450">https://bugs.webkit.org/show_bug.cgi?id=267450</a>
<a href="https://rdar.apple.com/120895293">rdar://120895293</a>

Reviewed by Aditya Keerthi.

Reuse the boilerplate zooming code from other controls for the switch
control so it works correctly when zoomed using &quot;Zoom In&quot;.

In SwitchTrackMac this scales both the inflatedTrackRect and the
trackRect as the latter is used for drawing the focus ring.

In SwitchThumbMac this adds a CGContextStateSaver for the
trackBuffer context for consistency with SwitchTrackMac. It isn&apos;t
required to address this bug. It also renames a variable for internal
consistency.

* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-expected.html: Added.
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-vertical-lr-rtl-expected.html: Added.
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-vertical-lr-rtl.html: Added.
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-rtl-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-rtl.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-lr-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-lr.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-rl-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-rl.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform.html: Added.
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::SwitchTrackMac::draw):

Canonical link: <a href="https://commits.webkit.org/273089@main">https://commits.webkit.org/273089@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83be96b3227ada7026a1a2f244ab854e97e86ad2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13026 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36206 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36906 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31002 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35301 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15412 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10156 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30034 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34729 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11028 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9628 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38197 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31092 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30886 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35821 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9858 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7724 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33736 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11644 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30187 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7879 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10417 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10676 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
```diff
@@ -89,15 +96,17 @@
 
     auto cgContext = trackBuffer->context().platformContext();
 
-    GraphicsContextStateSaver stateSaver(context);
```
I don't think we should be removing this, there's still `context.drawConsumingImageBuffer` below. No guarantee that doesn't modify context state.

annevk replied:
I'm not sure what you mean. I moved it up so the scale change ends up being in scope.

pxlcoder replied:
Sorry, I misread the diff.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
```diff
@@ -179,8 +187,6 @@
         return trackImage(context, maskImage, inflatedTrackRect.size(), deviceScaleFactor, style, isOn, isRTL, isVertical, isEnabled, isPressed, isInActiveWindow, needsOnOffLabels, coreUISize);
     };
 
-    GraphicsContextStateSaver stateSaver(context);
```
Ditto about not removing this.

annevk replied:
Ditto.

pxlcoder reviewed (commented):
+1 for tests.

nt1m commented on LayoutTests/fast/forms/switch/no-pixels-outside-the-box-expected.html:
```diff
@@ -0,0 +1,20 @@
+input, span { grid-area: 1/1 }
+</style>
+
+<input type=checkbox switch>
```
I would remove this from the ref?

annevk replied:
It's important to determine the size.

nt1m replied:
I wonder if we could just use a white overlay and have the ref be a blank file?

annevk replied:
Nice, will do that.

annevk committed:
Make <input type=checkbox switch> zoomable on macOS
https://bugs.webkit.org/show_bug.cgi?id=267450
rdar://120895293

Reviewed by Aditya Keerthi.

Reuse the boilerplate zooming code from other controls for the switch
control so it works correctly when zoomed using "Zoom In".

In SwitchTrackMac this scales both the inflatedTrackRect and the
trackRect as the latter is used for drawing the focus ring.

In SwitchThumbMac this adds a CGContextStateSaver for the
trackBuffer context for consistency with SwitchTrackMac. It isn't
required to address this bug. It also renames a variable for internal
consistency.

* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-expected.html: Added.
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-vertical-lr-rtl-expected.html: Added.
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box-vertical-lr-rtl.html: Added.
* LayoutTests/fast/forms/switch/no-pixels-outside-the-box.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-rtl-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-rtl.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-lr-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-lr.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-rl-expected.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform-vertical-rl.html: Added.
* LayoutTests/fast/forms/switch/zoom-approximates-transform.html: Added.
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
(WebCore::SwitchThumbMac::draw):
* Source/WebCore/platform/graphics/mac/controls/SwitchTrackMac.mm:
(WebCore::SwitchTrackMac::draw):

Canonical link: https://commits.webkit.org/273089@main
